### PR TITLE
Thepbjain/clean routes

### PR DESCRIPTION
**Pull request by ThePBJain**

### CODE REVIEW

File: src/server/routes/api/rfid.js [hunk 1]
```diff
@@ -2,8 +2,7 @@
  * Created by PranavJain on 2/20/17.
  */
 let express = require('express');
-let router = express.Router();
-
+let router = express.Router({});
```
Comment (sushshring) [inline src/server/routes/api/rfid.js]: Why is this necessary?

File: src/server/routes/api/rfid.js [hunk 2]
```diff
@@ -633,82 +443,112 @@ router.post('/add', helpers.ensureScannerAuthenticated, function (req, res, next
...
+            console.log("UNSENT SCANS: " + JSON.stringify(options.body));
+            //todo:worry about promises and mutli data usage
+            /*
+              200 if everything is success
+              207 if partial failure
+                - ones that succeeded will have the original scan
+                - ones that failed, it will contain an error with a status code
+                (409 = DUPLICATE or invalid relation) (500 for other issues)
+              400 if formatting of sent info is bad
+              409 if Duplicate
```
Comment (mdh5389) [inline src/server/routes/api/rfid.js]: Is this also "invalid relation"?

File: src/server/routes/api/rfid.js [hunk 3]
```diff
@@ -633,82 +443,112 @@ router.post('/add', helpers.ensureScannerAuthenticated, function (req, res, next
...
+              500 if everything failed
```
Comment (mdh5389) [inline src/server/routes/api/rfid.js]: You previously said 500 for other issues which is different than everything failing

File: src/server/routes/api/rfid.js [hunk 4]
```diff
@@ -860,7 +700,7 @@ router.get('/active-locations', function (req, res, next) {
 
 });
 
-
+/*
```
Comment (mdh5389) [inline src/server/routes/api/rfid.js]: Is this no longer needed? If so, why not just delete?

File: README.md [hunk 5]
```diff
@@ -39,6 +39,11 @@ docker-compose start
 docker-compose stop
 ```
 
+To bring everything down and remove the containers, use:
```
Comment (mdh5389) [inline README.md]: Thank you for adding this

File: src/server/routes/api/rfid.js [hunk 6]
```diff
@@ -52,38 +51,21 @@ let unsent_assignments = [];
...
+ * @api {post} /rfid/assignment Register RFID Band to User
  * @apiVersion 1.0.0
  * @apiName Register RFID
  * @apiGroup RFID
  * @apiDescription
  * Register RFID Band to User. Sends assignment to main server, while locally replacing user key to RFID code.
  * @apiPermission Scanner
  *
- * @apiParam {Number} id  RFID code to set to user.
+ * @apiParam {Number} rfid  RFID code to set to user.
```
[the remark this answers is not in the document]
Comment (mdh5389) [reply src/server/routes/api/rfid.js]: I don't know if we want this to be "rfid". Might be better begin "wid" for Wristband ID.

[the remark this answers is not in the document]
Comment (ThePBJain) [reply src/server/routes/api/rfid.js]: @sushshring thoughts?

[the remark this answers is not in the document]
Comment (sushshring) [reply src/server/routes/api/rfid.js]: Yeah makes sense to me. Seems like a much better name

File: src/server/routes/api/rfid.js (hunk 1 above)
Comment (ThePBJain) [reply src/server/routes/api/rfid.js]: no fundamental changes but is more expressive for what is expected.

File: src/server/routes/api/rfid.js (hunk 2 above)
Comment (ThePBJain) [reply src/server/routes/api/rfid.js]: yea it is, ill fix it

File: src/server/bin/www [hunk 7]
```diff
@@ -6,35 +6,39 @@
...
- */
+let server = null;
+if(process.env.USE_HTTPS){
+  /**
+   * Create HTTPS server.
+   */
+  var options = {
+    key: fs.readFileSync('./src/server/bin/keys/key.pem'),
+    cert: fs.readFileSync('./src/server/bin/keys/cert.pem'),
+    passphrase: "hackDaddy",
```
Comment (sushshring) [inline src/server/bin/www]: hardcoded?

Comment (ThePBJain) [reply src/server/bin/www]: idk why the paths work like this but they do.

Comment (sushshring) [reply src/server/bin/www]: I meant the passphrase

Comment (ThePBJain) [reply src/server/bin/www]: oh yea, the passphrase for the cert isn't typically necessary

File: src/server/routes/auth.js [hunk 8]
```diff
@@ -611,7 +628,7 @@ router.get('/removeall', helpers.ensureAdminJSON, function (req, res, next) {
 });
 
 /**
- * @api {get} /auth/removeall Remove all Scanners
+ * @api {get} /auth/scanner/removeall Remove all Scanners
```
Comment (sushshring) [inline src/server/routes/auth.js]: Thanks for this change

**Comment by ThePBJain** [comment]
> LGTM. I couldn't see any tests yet, but I assume that's an upcoming commit?

lol I thought I committed that last night. Guess I didn't

File: src/server/routes/auth.js [hunk 9]
```diff
@@ -273,12 +273,11 @@ router.get('/updatedb', helpers.ensureAdminJSON, function (req, res, next) {
     .then(function (response) {
       // Request was successful, use the response object at will
       //do redis stuff then
-      //todo: this is being treated synchronously when it's not synchronous fix with promises
       let numErrors = 0;
       let promises = [];
       //code to build promises to run
       response.map(function (element) {
-        console.log(element.rfid_uid);
+        //console.log(element.rfid_uid);
```
Comment (sushshring) [inline src/server/routes/auth.js]: As a backlog item, perhaps we could implement a simple logger that supports multiple levels of logging. Keeping these as debug calls, on production environments it would log only higher priority calls.

Comment (ThePBJain) [reply src/server/routes/auth.js]: Could you make an issue for this. It would help to keep track of future changes.

File: src/server/bin/keys/HTTPS.md [hunk 10]
```diff
@@ -0,0 +1,24 @@
...
+    openssl req -in server.csr -text -noout | grep -i "Signature.*SHA256" && echo "All is well" || echo "This certificate doesn't work in 2017! You must update OpenSSL to generate a widely-compatible certificate"
+
+    ```
+1. Copy the cert.pem and key.pem files to the `src/server/bin/keys/` directory.
+1. Open the `src/server/bin/www` file.
+    1. Uncomment the `Create HTTPS server.` section
+    1. Comment the `Create HTTP server.` section
+1. Get the fingerprint of the certificate: `openssl x509 -noout -in cert.pem -fingerprint`
+1. Save the fingerprint for verifying Redis identity on Scanners.
+TODO: make script for this
```
Comment (mdh5389) [inline src/server/bin/keys/HTTPS.md]: Seems easy enough to make this script.. why no just do it?

File: src/server/routes/auth.js (hunk 9 above)
Comment (sushshring) [reply src/server/routes/auth.js]: Issue on Github? or Asana?

Comment (ThePBJain) [reply src/server/routes/auth.js]: github

File: src/server/bin/www [hunk 11]
```diff
@@ -6,35 +6,40 @@
...
-  rejectUnauthorized: false
-};
-
-var server = https.createServer(options, app);
-*/
-/**
- * Create HTTP server.
- */
+let server = null;
+//TODO: fix flag
```
Comment (mdh5389) [inline src/server/bin/www]: What needs to be fixed?  Can it be fixed now?

File: src/server/routes/api/rfid.js [hunk 12]
```diff
@@ -52,38 +51,21 @@ let unsent_assignments = [];
...
-/* RESPONSE
-{
-    status: 'success',
-    data: "OK",
-    message: 'Created tab.'
-}
- */
 /**
- * @api {post} /tabs/setup Register RFID Band to User
+ * @api {post} /rfid/assignment Register Wristband to User
```
Comment (mdh5389) [inline src/server/routes/api/rfid.js]: Shouldn't these api routes be versioned as well?

File: test/routes/scanner.js [hunk 13]
```diff
@@ -0,0 +1,170 @@
...
+const chai = require('chai');
+const chaiHttp = require('chai-http');
+const app = require('../../src/server/app');
+const Scanner = require('../../src/server/models/scanner');
+let User = require('../../src/server/models/user');
+const should = chai.should();
+
+chai.use(chaiHttp);
+
+describe('api key test', () => {
```
Comment (sushshring) [inline test/routes/scanner.js]: One of the standards I've been following is prefacing the tests with the kind of test it is. For example, this would be:
`INTEGRATION TEST: GET /auth/scanner/register` and then the `it()` call actually describes what the test is expected to do.

File: test/routes/scanner.js [hunk 14]
```diff
@@ -0,0 +1,170 @@
...
+    chai.request(app)
+      .post('/auth/scanner/register')
+      .send(body)
+      .end((err, res) => {
+        should.equal(err, null);
+        res.should.have.status(200);
+        res.body.should.contain.keys(["status", "data", "message"]);
+        res.body.status.should.be.equal("success");
+        res.body.data.should.contain.keys(["apikey", "name"]);
+        //console.log(res.body);
```
Comment (sushshring) [inline test/routes/scanner.js]: Is the pin generated randomly? Is there any way you could seed the RNG and verify it is the expected apikey?

File: test/routes/scanner.js [hunk 15]
```diff
@@ -0,0 +1,170 @@
...
+        done();
+      });
+  });
+});
+describe('user auth test', () => {
```
Comment (sushshring) [inline test/routes/scanner.js]: nitpick: add whitespace between tests

File: test/routes/scanner.js [hunk 16]
```diff
@@ -0,0 +1,170 @@
...
+            console.log("AGENT BODY: " + res.body);
+            res.should.have.status(200);
+            agent.close();
+            done();
+          });
+      });
+
+  });
+
+  it('it should get api key for scanner', (done) => {
```
Comment (sushshring) [inline test/routes/scanner.js]: Is this the same test as above?

File: src/server/routes/api/rfid.js [hunk 17]
```diff
@@ -408,221 +367,73 @@ router.post('/getpin', helpers.ensureScannerAuthenticated, function (req, res, n
...
+                message: 'Something went wrong'
               });
-
-              console.log("Incrementing numScans counter");
-              if (obj) {
-
+          } else {
+            console.log("Incrementing "+incrementedKey+" counter");
+            if (obj) {
+              if (incrementedKey === "numScans"){
```
Comment (mdh5389) [inline src/server/routes/api/rfid.js]: You can add this condition to the if statement above this line unless there is a reason to keep them separate

File: test/routes/scanner.js (hunk 14 above)
Comment (ThePBJain) [reply test/routes/scanner.js]: its randomly generated. if it returns this response it should work unless there some crazy bug in the dependency...

File: test/routes/scanner.js [hunk 18]
```diff
@@ -0,0 +1,170 @@
...
+
+//TODO: figure out how to reset redis to disassociate scans (maybe an after tag that unassociates them
+describe('scanner proper flow test', () => {
+  let apikey = "";
+  //part of test data on redis
+  let pin = 512;
+  let wristbandID = "TEST_WID";
+  let agent = chai.request.agent(app);
+
+  before(function (done){
```
Comment (sushshring) [inline test/routes/scanner.js]: Can you add some explanation why this `before()` is necessary?

File: test/routes/scanner.js [hunk 19]
```diff
@@ -0,0 +1,170 @@
...
+      "pin": pin,
+      "apikey": apikey
+    };
+    chai.request(app)
+      .post('/rfid/getpin')
+      .send(body)
+      .end((err, res) => {
+        should.equal(err, null);
+        res.should.have.status(200);
+        res.body.should.contain.keys(["status", "data", "message"]);
```
Comment (sushshring) [inline test/routes/scanner.js]: Can you make any assertions on the data?

File: test/routes/scanner.js [hunk 20]
```diff
@@ -0,0 +1,170 @@
...
+      });
+  });
+  it('it should associate wristband id to user', (done) => {
+    let body = {
+      "wid": wristbandID,
+      "pin": pin,
+      "apikey": apikey
+    };
+    chai.request(app)
+      .post('/rfid/assignment')
```
Comment (sushshring) [inline test/routes/scanner.js]: This assignment will work the first time you insert it. As long as Redis isn't going to SQL, this should be okay. Otherwise it will fail with a 400 the second time these tests are run.

File: test/routes/scanner.js (hunk 18 above)
Comment (ThePBJain) [reply test/routes/scanner.js]: done

File: src/server/routes/api/rfid.js (hunk 12 above)
Comment (ThePBJain) [reply src/server/routes/api/rfid.js]: @sushshring how you version?

Comment (sushshring) [reply src/server/routes/api/rfid.js]: Currently I just have a `/v1/` prepended to all routes. If we end up supporting multiple versions, it'll go under a subfolder and be put under something like a `v2`

File: test/routes/scanner.js (hunk 20 above)
Comment (ThePBJain) [reply test/routes/scanner.js]: it will still return 200 to the scanner or test case as it handless the 400 in the backend

Comment (ThePBJain) [reply test/routes/scanner.js]: it technically isn't a failed test but I see your point. How should we collect that information though since it's not returning those 400 errors.

File: test/routes/scanner.js (hunk 19 above)
Comment (ThePBJain) [reply test/routes/scanner.js]: done

File: src/server/bin/keys/disableHTTPS.sh [hunk 21]
```diff
@@ -0,0 +1,4 @@
+#!/bin/bash
+#update .env flag to disable https
+sed -i "" 's/true/false/g' ../../../../.env
```
Comment (sushshring) [inline src/server/bin/keys/disableHTTPS.sh]: Should you check for the HTTPS_ENABLED as well? this will replace all trues to false

File: src/server/routes/api/rfid.js (hunk 17 above)
Comment (ThePBJain) [reply src/server/routes/api/rfid.js]: There's a scenario where incrementedKey is not "numScans" and I want it to still go in the 'obj' if statement

File: src/server/routes/api/rfid.js (hunk 12 above)
Comment (ThePBJain) [reply src/server/routes/api/rfid.js]: just set up versioning

File: src/server/bin/keys/disableHTTPS.sh (hunk 21 above)
Comment (ThePBJain) [reply src/server/bin/keys/disableHTTPS.sh]: fixed

File: apidoc.json [hunk 22]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "Redis API Documentation",
-  "version": "1.0.0",
+  "version": "1.5.0",
```
Comment (sushshring) [inline apidoc.json]: ? seems arbitrary?

Comment (ThePBJain) [reply apidoc.json]: it is half way point before v2

Comment (mdh5389) [inline apidoc.json]: How did we get to 1.5.0.  Isn't versioning usually incrementally W.R.T. master so it would be 1.1.0

Comment (sushshring) [reply apidoc.json]: It does not follow semantic versioning. If you’re changing a version, make it 1.1

File: src/server/bin/keys/enableHTTPS.sh [hunk 23]
```diff
@@ -0,0 +1,22 @@
...
+openssl genrsa -des3 -out key.pem 2048
+
+# create certificate signing request, enter "*" as a "Common Name", leave "challenge password" blank
+openssl req -new -sha256 -key key.pem -out server.csr
+
+# generate self-signed certificate for 1 year
+openssl req -x509 -sha256 -days 365 -key key.pem -in server.csr -out cert.pem
+
+# validate the certificate
+openssl req -in server.csr -text -noout | grep -i "Signature.*SHA256" && echo "All is well" || echo "This certificate doesn't work in 2017! You must update OpenSSL to generate a widely-compatible certificate"
```
[the remark this answers is not in the document]
Comment (sushshring) [reply src/server/bin/keys/enableHTTPS.sh]: Holy shit I didn't know about the suggested changes

File: src/server/routes/api/rfid.js [hunk 24]
```diff
@@ -181,6 +163,7 @@ router.post('/setup', helpers.ensureScannerAuthenticated, function (req, res, ne
                 let options = helpers.clone(serverOptions);
                 options.method = 'POST';
                 options.uri = options.uri + '/v1/scanner/assignment';
+                //TODO: Normalize with other sent scans... talk to sush
```
[the remark this answers is not in the document]
Comment (ThePBJain) [reply src/server/routes/api/rfid.js]: We need to talk about changing the variable names on the server.

File: apidoc.json (hunk 22 above)
Comment (ThePBJain) [reply apidoc.json]: we'll switch this to 2.0
